driver-core/src/main/java/com/datastax/driver/core/ConvictionPolicy.java

```diff
@@ -50,30 +70,90 @@
          * @param host the host this policy applies to
          * @return the newly created {@link ConvictionPolicy} instance.
          */
-        public ConvictionPolicy create(Host host);
+        ConvictionPolicy create(Host host, ReconnectionPolicy reconnectionPolicy);
     }
 
-    public static class Simple extends ConvictionPolicy {
-        private Simple(Host host) {
+    static class DefaultConvictionPolicy extends ConvictionPolicy {
+        private final Host host;
+        private final ReconnectionPolicy reconnectionPolicy;
+        private final AtomicInteger openConnections = new AtomicInteger();
+
+        private volatile long nextReconnectionTime = Long.MIN_VALUE;
+        private ReconnectionPolicy.ReconnectionSchedule reconnectionSchedule;
+
+        private DefaultConvictionPolicy(Host host, ReconnectionPolicy reconnectionPolicy) {
+            this.host = host;
+            this.reconnectionPolicy = reconnectionPolicy;
         }
 
         @Override
-        public boolean addFailure(ConnectionException exception) {
-            return true;
+        void signalConnectionCreated() {
+            int newCount = openConnections.incrementAndGet();
+            Host.statesLogger.debug("[{}] new connection created, total = {}", host, newCount);
+            resetReconnectionTime();
         }
 
-        public boolean addFailureFromExternalDetector() {
-            return true;
+        @Override
+        void signalConnectionClosed() {
+            if (host.state == Host.State.DOWN)
+                return;
+            int remaining = openConnections.decrementAndGet();
+            assert remaining >= 0;
```

Comment:
I noticed a valid case for this:   If someone calls Cluster.close() it ultimately calls Connection.closeAsync() which calls signalConnectionClosed(), depending on ordering of connections closing you may sometimes get this exception:

```
java.lang.AssertionError
	at com.datastax.driver.core.ConvictionPolicy$DefaultConvictionPolicy.signalConnectionClosed(ConvictionPolicy.java:101)
	at com.datastax.driver.core.Connection.closeAsync(Connection.java:622)
	at com.datastax.driver.core.HostConnectionPool.discardAvailableConnections(HostConnectionPool.java:607)
	at com.datastax.driver.core.HostConnectionPool.closeAsync(HostConnectionPool.java:585)
	at com.datastax.driver.core.SessionManager.closeAsync(SessionManager.java:116)
	at com.datastax.driver.core.Cluster$Manager.close(Cluster.java:1513)
	at com.datastax.driver.core.Cluster$Manager.access$200(Cluster.java:1199)
	at com.datastax.driver.core.Cluster.closeAsync(Cluster.java:490)
	at com.datastax.driver.core.Cluster.close(Cluster.java:501)
	at com.datastax.driver.core.HostConnectionPoolTest.should_not_close_trashed_connection_until_no_in_flight(HostConnectionPoolTest.java:335)
```